rt/x86-64-linux/System.v3

```diff
@@ -8,7 +8,120 @@
+ * Inet6 for IPv6 addresses.
+ */
+type InetAddress {
+	case Inet4(addr: Array<u8>);
+	case Inet6(addr: Array<u8>);
+}
+
+class SocketBase {
+	// IP address of socket
+	def addr: InetAddress;
+	// Associated file descriptor of socket
+	var fd: int;
+	// Local port the socket is connected to
+	def localPort: int;
+	// Server port the socket is connected to
+	def port: int;
+
+	new(addr, localPort, port) {
+		var family: int;
+		match (addr) {
+			Inet4(addrV4) => family = LinuxConst.AF_INET;
+			Inet6(addrV6) => family = LinuxConst.AF_INET6;
+		}
+		fd = System.socket(family, LinuxConst.SOCK_STREAM /* type */, 0 /* protocol */);
+	}
+}
+
+class ClientSocket extends SocketBase {
+	new(addr: InetAddress, localPort: int, port: int) super(addr, localPort, port) { }
+
+	def connect() -> int {
+		match (addr) {
+			Inet4(addrV4) => {
+				// Definition of sockaddr for IPv4:
+				// struct sockaddr_in {
+				//   short int          sin_family;  // Address family, AF_INET
+				//   unsigned short int sin_port;    // Port number
+				//   struct in_addr     sin_addr;    // Internet address
+				//   unsigned char      sin_zero[8]; // Same size as struct sockaddr
+				// };
+				def sockAddr = Array<byte>.new(LinuxConst.SOCKADDR_V4_SIZE);
+				def nboPort = u16ToNetworkByteOrder(u16.view(port));
+				def nboAddr = u32ToNetworkByteOrder(
+						u32.view((int.view(addrV4[0]) << 24) | (int.view(addrV4[1]) << 16)
+							| (int.view(addrV4[2]) << 8) | addrV4[3])
+				);
+				// XXX: Disgusting hack alert! Hopefully this can be cleaned up with structs
+				sockAddr[0] = byte.view(LinuxConst.AF_INET);
+				sockAddr[1] = 0x00;
+				sockAddr[2] = byte.view(nboPort & 0xFF);
+				sockAddr[3] = byte.view((nboPort >> 8)  & 0xFF);
+				sockAddr[4] = byte.view(nboAddr & 0xFF);
+				sockAddr[5] = byte.view((nboAddr >> 8)  & 0xFF);
+				sockAddr[6] = byte.view((nboAddr >> 16) & 0xFF);
+				sockAddr[7] = byte.view((nboAddr >> 24) & 0xFF);
+				return System.connect(fd, sockAddr, LinuxConst.SOCKADDR_V4_SIZE);
+			}
+			Inet6(addrV6) => System.error("Unimplemented", "IPv6 connections are currently unsupported!");
+		}
+
+		return -1;
+	}
+
+	def read(buf: Array<byte>) -> int {
+		return System.fileReadK(fd, buf, 0, buf.length);
+	}
+
+	def write(msg: string) {
+		System.fileWriteK(fd, msg, 0, msg.length);
+	}
+
+	def close() {
+		System.fileClose(fd);
+	}
+}
+
+// class ServerSocket extends SocketBase {
+//
+// }
+
+// Network Byte Order is defined to be big-endian. These functions assume
+// little-endian architecture. TODO: Have to make this a nop for big-endian architectures
+def u16ToNetworkByteOrder(n: u16) -> u16 {
+	return u16.view(((n & 0xFF) << 8) | ((n & 0xFF00) >> 8));
+}
+
+def u32ToNetworkByteOrder(n: u32) -> u32 {
+	return u32.view(((n & 0xFF) << 24) | ((n & 0xFF00) << 8) |
+		((n & 0xFF0000) >> 8) | ((n & 0xFF000000) >> 24));
+}
+
 component System {
+	def ANY_V4 = InetAddress.Inet4([0, 0, 0, 0]);
```

Comment:
I think it'd be best not to add anything to `System` just yet. So far it's just the minimal portable stuff to get off the ground. It feels like this should probably live under `lib/net` for now. It's platform-specific, so maybe `lib/net/x86-64-linux` until we can figure out a nicer way to make portable interfaces that automatically build against the right target library.

Reply:
Including the system calls like `socket`?

Reply:
I've moved all the syscalls to `lib/net/x86-64-linux/Net.v3`.